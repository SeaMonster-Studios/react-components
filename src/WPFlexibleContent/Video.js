import React from 'react'
import PropTypes from 'prop-types'
import sanitizeHtml from 'sanitize-html'

export const Video = props => <span>item</span>

Video.propTypes = {}
