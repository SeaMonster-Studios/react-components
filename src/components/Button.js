/**
 * @name Button
 * @example
 *  <Button baseColor='rgb(255,127,104)' textColor='rgb(255,255,255)' hoverStyle='ripple'>Contact Us</Button>
 */

import * as React from "react"
import PropTypes from "prop-types"
import styled, { keyframes, css } from "react-emotion"

const Wrapper = styled("button")`
  text-decoration: none;
  border-color: ${(props) => props.borderColor};
  cursor: pointer;

  ${(props) => colorProps(props)};
  ${(props) => hoverProps(props)};

  ${(props) => props.styles};
`

function colorProps(props) {
  if (props.inverse) {
    switch (props.inverseStyle) {
      case "transparent":
        return css`
          background-color: rgba(255, 255, 255, 0);
          color: ${props.baseColor};
        `
      case "default":
      default:
        return css`
          background-color: ${props.textColor};
          color: ${props.baseColor};
        `
    }
  } else {
    return css`
      background-color: ${props.baseColor};
      color: ${props.textColor};
    `
  }
}

const rippleOut = keyframes`
  100% {
    top: -12px;
    right: -12px;
    bottom: -12px;
    left: -12px;
    opacity: 0;
  }
`

function hoverProps(props) {
  switch (props.hoverStyle) {
    case "ripple":
      return css`
        transform-origin: center;
        position: relative;
        border-color: ${props.hoverBorderColor};

        &::after {
          content: "";
          position: absolute;
          border: 1px solid;
          border-radius: inherit;
          border-color: inherit;
          top: -1px;
          right: -1px;
          bottom: -1px;
          left: -1px;
        }

        &:hover,
        &:focus {
          text-decoration: none;
          color: ${props.inverse ? props.hoverBaseColor : props.hoverTextColor};
          background-color: ${props.inverse
            ? props.hoverTextColor
            : props.hoverBaseColor};

          &::after {
            animation: ${rippleOut} 0.5s;
          }
        }
      `
    case "default":
    default:
      return css`
        text-decoration: none;

        &:hover,
        &:focus {
          color: ${props.inverse ? props.hoverBaseColor : props.hoverTextColor};
          background-color: ${props.inverse
            ? props.hoverTextColor
            : props.hoverBaseColor};

          border-color: ${props.hoverBorderColor};
        }
      `
  }
}

export const Button = (props) => {
  const {
    baseColor,
    textColor,
    borderColor,
    hoverBaseColor,
    hoverTextColor,
    hoverBorderColor,
    inverse,
    inverseStyle,
    hoverStyle,
    style,
    className,
    children,
  } = props

  return (
    <Wrapper
      {...{
        "data-testid": "component-button",
        baseColor,
        textColor,
        borderColor: borderColor || baseColor,
        inverse: inverse,
        inverseStyle: inverseStyle,
        hoverStyle: hoverStyle,
        hoverBaseColor: hoverBaseColor || baseColor,
        hoverTextColor: hoverTextColor || textColor,
        hoverBorderColor: hoverBorderColor || hoverBaseColor || baseColor,
        style,
        className,
      }}
    >
      {children}
    </Wrapper>
  )
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  /** Css color value. */
  baseColor: PropTypes.string,
  /** Css color value. */
  textColor: PropTypes.string,
  /** Css color value. Defaults to baseColor */
  borderColor: PropTypes.string,
  /** Css color value. Defaults to baseColor */
  hoverBaseColor: PropTypes.string,
  /** Css color value. Defaults to textColor */
  hoverTextColor: PropTypes.string,
  /** Css color value. Defaults to hoverBaseColor || baseColor */
  hoverBorderColor: PropTypes.string,
  inverse: PropTypes.bool,
  inverseStyle: PropTypes.oneOf(["default", "transparent"]),
  hoverStyle: PropTypes.oneOf(["default", "ripple"]),
  className: PropTypes.string,
  style: PropTypes.object,
}

Button.defaultProps = {
  inverse: false,
  inverseStyle: "default",
  hoverStyle: "default",
  textColor: "rgb(255,255,255)",
  baseColor: "rgb(0,0,0)",
  className: "",
  style: {
    transition: "all 0.5s ease",
    padding: "10px 30px",
  },
}
