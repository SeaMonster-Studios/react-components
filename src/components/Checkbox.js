import * as React from "react"
import PropTypes from "prop-types"
import { Transition } from "react-spring"
import styled, { css } from "react-emotion"

const borderWidth = 2

export const Wrapper = styled("button")`
  width: ${(props) => props.size}px;
  height: ${(props) => props.size}px;
  border: ${borderWidth}px solid #000;
  cursor: pointer;
  transition: all 0.5s ease;
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;

  .input-wrapper {
    min-width: ${(props) => props.size}px;
    min-height: ${(props) => props.size}px;
    margin-top: -${borderWidth * 2}px;
    overflow: hidden;
    position: relative;
    padding: 0;
  }

  input {
    opacity: 0;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
    transform: scale(100);
    cursor: pointer;
    position: absolute;
    z-index: 2;
  }

  &:focus {
    outline: none;
  }

  .mark {
    position: absolute;
    z-index: 1;
    margin: auto;
    display: flex;
    align-items: center;
    justify-content: center;

    > * {
      position: absolute;
    }
  }

  ${(props) => props.styles} ${(props) => setActiveStyles(props)};
`

function setActiveStyles(props) {
  if (props.isChecked && !props.hasOnIcon) {
    return css`
      background-color: #000;

      ${props.activeStyles};
    `
  } else if (props.isChecked) {
    return props.activeStyles
  } else {
    return ""
  }
}


export class Checkbox extends React.Component {
  static propTypes = {
    /** Height and Width */
    size: PropTypes.number,
    className: PropTypes.string,
    style: PropTypes.object,
    activeStyles: PropTypes.string,
    /** For user/non-component controlled state */
    checked: PropTypes.bool,
    /** For user/non-component controlled state */
    onChange: PropTypes.func,
    /** Receives new value each time it's updated */
    checkedHasChanged: PropTypes.func,
    /** Svg (component) used when `checked` is true */
    OnIcon: PropTypes.func,
    /** Svg (component) used when `checked` is flase */
    OffIcon: PropTypes.func,
    iconsTransitionFrom: PropTypes.object,
    iconsTransitionEnter: PropTypes.object,
    iconsTransitionLeave: PropTypes.object,
  }
  static defaultProps = {
    size: 30,
    style: {},
    className: "",
    activeStyles: "",
    iconsTransitionFrom: {
      opacity: 0,
      transform: "rotate(90deg)",
    },
    iconsTransitionEnter: {
      opacity: 1,
      transform: "rotate(0deg)",
    },
    iconsTransitionLeave: {
      opacity: 0,
      transform: "rotate(-90deg)",
    },
  }
  state = {
    checked: false,
  }
  constructor(props) {
    super(props)

    if (
      (this.props.checked !== undefined && !this.props.onChange) ||
      (this.props.checked === undefined && this.props.onChange)
    )
      throw new Error(
        "If you want to control the state of Checkbox, you must provide both the `checked` and `onChange` props.",
      )
  }
  handleChange = () => {
    this.setState((prevState) => {
      const checked = !prevState.checked

      if (this.props.checkedHasChanged) this.props.checkedHasChanged(checked)

      return { checked }
    })
  }
  render() {
    const {
      size,
      style,
      className,
      activeStyles,
      checked,
      onChange,
      // eslint-disable-next-line no-unused-vars
      checkedHasChanged, // destruct so it's not in the attrs object
      OffIcon,
      OnIcon,
      iconsTransitionFrom,
      iconsTransitionEnter,
      iconsTransitionLeave,
      ...attrs
    } = this.props

    const isChecked = checked || this.state.checked

    return (
      <Wrapper
        {...{
          "data-testid": "component-checkbox",
          className,
          isChecked,
          size,
          hasOnIcon: OnIcon ? true : false,
          hasOffIcon: OffIcon ? true : false,
          style: style,
          activeStyles: activeStyles,
          ...attrs,
        }}
      >
        <div
          className="input-wrapper"
          data-testid="component-checkbox-input-wrapper"
        >
          <input
            type="checkbox"
            data-testid="component-checkbox-input"
            onChange={onChange || this.handleChange}
            checked={isChecked}
          />
        </div>

        {OnIcon &&
          OffIcon && (
            <Transition
              from={iconsTransitionFrom}
              enter={iconsTransitionEnter}
              leave={iconsTransitionLeave}
            >
              {isChecked
                ? (styles) => buildIcon(OnIcon, styles, "on-mark")
                : (styles) => buildIcon(OffIcon, styles, "off-mark")}
            </Transition>
          )}
      </Wrapper>
    )
  }
}

const buildIcon = (Icon, styles, testid) => (
  <div style={styles} className="mark" data-testid={testid}>
    <Icon />
  </div>
)
