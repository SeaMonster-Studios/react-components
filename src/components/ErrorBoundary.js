import * as React from "react"
import PropTypes from "prop-types"
import * as Sentry from "@sentry/browser"
import styled from "react-emotion"

const Wrapper = styled("div")`
  background-color: #f8ebe6;
  box-shadow: inset 0 0 0 0 transparent, 0 0 0 1px rgba(63, 63, 68, 0.05),
    0 1px 3px 0 rgba(63, 63, 68, 0.15);
  position: absolute;
  border-top: 2px solid rgb(222, 54, 24);
  top: 50%;

  width: calc(100% - 40px);
  margin: 0 20px;
  padding: 20px;
  text-align: center;
  line-height: 1.5;

  &,
  button {
    font-size: 1rem;
    font-weight: 300;
  }

  h2 {
    font-size: 1.625rem;
    padding-bottom: 10;
  }

  button {
    border: none;
    background: none;
    cursor: pointer;
    padding: 0;
    border-bottom: 1px solid rgb(222, 54, 24);
  }
`

export class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    style: PropTypes.object,
    className: PropTypes.string,
  }
  static defaultProps = {
    style: {},
    className: "",
  }
  state = {
    hasError: false,
  }

  componentDidCatch(error, errorInfo) {
    this.setState(() => {
      return {
        hasError: true,
      }
    })

    if (
      process.env.NODE_ENV !== "development" &&
      typeof document !== "undefined" &&
      error.message != "IDontExist is not defined"
    ) {
      /* eslint-disable-next-line no-console */
      console.error("Errors sent to Sentry.io", error, errorInfo)

      Sentry.configureScope((scope) => {
        Object.keys(errorInfo).forEach((key) => {
          scope.setExtra(key, errorInfo[key])
        })
      })
      Sentry.captureException(error)

    } else {
      /* eslint-disable-next-line no-console */
      console.error(
        "Error caught in Error ErrorBoundary. This will reported to Sentry when not in development.",
      )
    }
  }

  render() {
    const { children, style, className } = this.props
    if (this.state.hasError) {
      return (
        <Wrapper
          data-testid="component-error-boundary"
          className={className}
          style={style}
        >
          <div>
            <h2>Sorry, something went wrong.</h2>
            <ReportForm />
            <p>Please try reloading the page.</p>
          </div>
        </Wrapper>
      )
    }
    return children
  }
}

const ReportForm = () =>
  process.env.NODE_ENV !== "development" && typeof document !== "undefined" ? (
    <Form />
  ) : (
    <span />
  )

const Form = () => (
  <p data-testid="component-error-boundary-form">
    This error has been reported to our development team. Please{" "}
    <button onClick={() => Sentry.showReportDialog()}>
      click here to fill out a report.
    </button>
  </p>
)
