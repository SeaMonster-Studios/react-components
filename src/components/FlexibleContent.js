import React from "react"
import PropTypes from "prop-types"
import { Trail, animated } from "react-spring"
import Component from "@reactions/component"
import { OneColumn } from "./OneColumn"
import { TwoColumns } from "./TwoColumns"
import { TwoColumnsImageGrid } from "./TwoColumnsImageGrid"
import { TwoColumnsImageLeads } from "./TwoColumnsImageLeads"
import { TwoColumnsStacked } from "./TwoColumnsStacked"
import { TwoColumnsThreeColumnList } from "./TwoColumnsThreeColumnList"
import { OneColumnVideo } from "./OneColumnVideo"
import { Grid123 } from "./Grid123"
import { wrapIframesInResponsiveVideo } from "../utils/index"

import styled, { css } from "react-emotion"

const common = (rowSpace, columnSpace) => css`
  overflow: hidden;
  padding-top: ${rowSpace}px;
  padding-bottom: ${rowSpace}px;

  .title,
  .subtitle {
    text-align: center;
  }

  .column .title,
  .column .subtitle {
    text-align: initial;
  }

  .title {
    margin-bottom: ${columnSpace}px;

    + .subtitle {
      margin-bottom: ${columnSpace}px;
      margin-top: -${columnSpace}px;
    }
  }

  .caption,
  figcaption {
    font-size: 14px;
    line-height: 1.125;
    background-color: #f2f2f2;
    padding: 10px;
  }

  img {
    max-width: 100%;
  }
  img {
    &:not(.thumbnail) {
      margin-bottom: ${columnSpace * 0.5}px;
    }

    + .caption {
      margin-top: -${columnSpace * 0.5}px;
    }

    &.alignleft {
      margin-right: ${columnSpace * 0.5}px;
    }

    &.alignright {
      margin-left: ${columnSpace * 0.5}px;
    }

    &.aligncenter {
      margin-left: auto;
      margin-right: auto;
      display: block;
    }
  }

  .column:not(:first-of-type) {
    margin-top: ${columnSpace}px;
  }

  .row,
  .column,
  .column-single {
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin-top: ${columnSpace}px;
    }
  }

  .row + .row {
    margin-top: ${columnSpace}px;
  }

  .column-grid {
    display: flex;
    flex-wrap: wrap;
    margin-left: -${columnSpace * 0.25}px;
    margin-right: -${columnSpace * 0.25}px;
    margin-bottom: -${columnSpace * 0.5}px;

    > div {
      width: calc(50% - ${columnSpace * 0.25}px);
      margin-bottom: ${columnSpace * 0.5}px;

      &:nth-of-type(odd) {
        padding-right: ${columnSpace * 0.25}px;
      }

      &:nth-of-type(even) {
        padding-left: ${columnSpace * 0.25}px;
      }

      img {
        width: 100%;
        display: block;
      }
    }
  }
`

export const Wrapper = styled("section")`
  ${(props) => css`
    ${common(props.rowSpace, props.columnSpace)};

    @media (min-width: ${props.breakpoint}px) {
      .column:not(:first-of-type) {
        margin-top: 0;
      }

      .column-grid {
        margin-left: ${props.columnSpace * 4}px;
      }

      img.column-lead {
        width: auto;
        max-width: 100%;
        min-height: ${props.minHeight || 0};
      }

      .row {
        display: flex;

        &:not(:last-of-type) {
          margin-bottom: ${props.columnSpace}px;
        }

        &.row-stacked {
          .column-image {
            margin-bottom: 0;
          }
        }

        + .row {
          margin-top: 0;
        }
      }

      .column-half {
        width: 50%;

        &:first-of-type:not(.order-bp-2),
        &.order-bp-1 {
          padding-right: ${props.columnSpace * 0.5}px;
        }

        &:last-of-type:not(.order-bp-1),
        &.order-bp-2 {
          padding-left: ${props.columnSpace * 0.5}px;
        }
      }

      .order-bp-2 {
        order: 2;
      }

      .order-bp-1 {
        order: 1;
      }

      .column-third {
        width: 33.333333333%;

        &:nth-of-type(3n - 2) {
          padding-right: ${props.columnSpace * 0.5}px;
        }

        &:nth-of-type(3n - 1) {
          padding-right: ${props.columnSpace * 0.5}px;
          padding-left: ${props.columnSpace * 0.5}px;
        }

        &:nth-of-type(3n) {
          padding-left: ${props.columnSpace * 0.5}px;
        }
      }

      .bp-align-center {
        align-items: center;
      }
    }
  `};
`

export const Grid123Wrapper = styled("section")`
  ${(props) => css`
    ${common(props.rowSpace, props.columnSpace)};

    .grid-item:not(:last-of-type) {
      margin-bottom: ${props.columnSpace}px;
    }

    @media (min-width: ${props.breakpoint1}px) {
      .grid {
        display: flex;
        width: 100%;
        flex-flow: row wrap;
      }

      .grid-item-type-image img {
        width: 100%;
        display: block;
      }

      .grid-item {
        width: calc(50% - 15px);
        margin: 0;

        &:nth-of-type(odd) {
          margin-right: 30px;
        }
      }
    }

    @media (min-width: ${props.breakpoint2}px) {
      .grid {
        display: flex;
        flex-flow: row wrap;
      }

      .grid-item {
        width: calc(33.333333333% - 20px);

        &:nth-of-type(odd) {
          margin-right: 0;
        }

        &:nth-of-type(3n - 2),
        &:nth-of-type(3n - 1) {
          margin-right: 30px;
        }
      }
    }
  `};
`

const layoutTypes = [
  "one_column",
  "two_columns",
  "two_columns_image_leads",
  "two_columns_stacked",
  "two_columns_three_column_list",
  "two_columns_image_grid",
  "video",
  "content_grid",
  "image_grid",
]

class FlexibleContentItem extends React.Component {
  static propTypes = {
    layout: PropTypes.oneOf(layoutTypes).isRequired,
    content: PropTypes.object,
    layoutProps: PropTypes.object,
  }
  render() {
    const { layout, layoutProps } = this.props
    switch (layout) {
      case "content_grid":
        return <Grid123 {...layoutProps} type="cms-content" />
      case "image_grid":
        return <Grid123 {...layoutProps} type="cms-images" />
      case "one_column":
        return <OneColumn {...layoutProps} />
      case "two_columns":
        return <TwoColumns {...layoutProps} />
      case "two_columns_image_leads":
        return <TwoColumnsImageLeads {...layoutProps} />
      case "two_columns_stacked":
        return <TwoColumnsStacked {...layoutProps} />
      case "two_columns_three_column_list":
        return <TwoColumnsThreeColumnList {...layoutProps} />
      case "two_columns_image_grid":
        return <TwoColumnsImageGrid {...layoutProps} />
      case "video":
        return <OneColumnVideo {...layoutProps} />
      default:
        return null
    }
  }
}

const AnimatedFlexibleContentItem = animated(FlexibleContentItem)

function getitemsProps(itemsProps, layout, animatedStyles) {
  const item = itemsProps.filter((item) => item.item === layout)[0]

  // There are default props for each item. However, if the user only passes in one prop for that layout type (say a className for one_column) but doesn't pass in the others (say style) then the defaultProp will be completely overridden and style wiill be undefined.

  if (item) {
    if (item.className && item.style) {
      return {
        ...item,
        className: `${layout} ${item.className} `,
        style: {},
      }
    } else if (!item.style) {
      return {
        ...item,
        lassName: `${layout} ${item.className} `,
        style: animatedStyles,
      }
    } else if (!item.className) {
      return {
        ...item,
        className: layout,
        style: {
          ...item.style,
          ...animatedStyles,
        },
      }
    } else {
      return { ...item, animatedStyles }
    }
  }
  return { ...item, style: animatedStyles }
}

export function FlexibleContent({
  className,
  items,
  style,
  rowSpace,
  columnSpace,
  breakpoint,
  itemsProps,
}) {
  return (
    <div
      data-testid="component-wp-flexible-content"
      style={style}
      className={`${className} flexible-content-component`}
    >
      <Component didMount={wrapIframesInResponsiveVideo}>
        <Trail
          from={{ opacity: 0, transform: "scale(0.99)" }}
          to={{ opacity: 1, transform: "scale(1)" }}
          keys={items.map(({ acf_fc_layout: layout }, i) => `${layout}-${i}`)}
        >
          {items.map(({ acf_fc_layout: layout, ...content }) => (styles) => (
            <AnimatedFlexibleContentItem
              {...{
                layout,
                layoutProps: {
                  ...content,
                  ...getitemsProps(itemsProps, layout, styles),
                  rowSpace,
                  columnSpace,
                  breakpoint,
                },
              }}
            />
          ))}
        </Trail>
      </Component>
    </div>
  )
}

FlexibleContent.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      acf_fc_layout: PropTypes.oneOf(layoutTypes).isRequired,
    }).isRequired,
  ).isRequired,
  className: PropTypes.string,
  style: PropTypes.object,
  /** Vertical spacing base */
  rowSpace: PropTypes.number,
  /** Horizontal spacing base */
  columnSpace: PropTypes.number,
  /** Mobile first breakpoint */
  breakpoint: PropTypes.number,
  /** pass props that will be applied to each item of the type provided. Layout Types: [
  'one_column',
  'two_columns',
  'two_columns_image_leads',
  'two_columns_stacked',
  'two_columns_three_column_list',
  'two_columns_image_grid',
  'video',
  'content_grid',
  'image_grid'
]*/
  itemsProps: PropTypes.arrayOf(
    PropTypes.shape({
      item: PropTypes.oneOf(layoutTypes).isRequired,
      className: PropTypes.string,
      style: PropTypes.object,
    }).isRequired,
  ),
}

FlexibleContent.defaultProps = {
  className: "",
  adminclass: "",
  style: {},
  rowSpace: 60,
  columnSpace: 30,
  breakpoint: 992,
  itemsProps: layoutTypes.map((type) => ({
    item: type,
    className: "",
    style: {},
  })),
}
