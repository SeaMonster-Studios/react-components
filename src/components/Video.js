import React from "react"
import PropTypes from "prop-types"
import { setHtml } from "../utils/index"
import { LazyLoadImage } from "./LazyLoadImage"
import { Play as PlayIcon } from "./Icons"
import styled, { injectGlobal } from "react-emotion"

injectGlobal(`
  .video-responsive {
    overflow: hidden;
    padding-bottom: 56.25%;
    position: relative;
    height: 0;
  }

  .video-responsive iframe {
    left: 0;
    top: 0;
    height: 100%;
    width: 100%;
    position: absolute;
  }
`)

export const Wrapper = styled("div")`
  .caption {
    font-size: 14px;
    line-height: 1.125;
    background-color: rgb(0, 0, 0, 0.05);
    padding: 10px;
  }

  .video-toggle {
    background: none;
    border: none;
    padding: 0;
    margin: 0;
    cursor: pointer;
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden;

    img {
      position: relative;
      z-index: 1;
      max-width: 100%;
    }

    .toggle-bg {
      height: 100%;
      width: 100%;
      background: rgba(0, 0, 0, 0.8);
      z-index: 2;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      transition: width 0.2s ease, height 0.2s ease, border 0.2s ease;
    }

    .icon-wrapper {
      z-index: 3;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      display: flex;
      align-items: center;
      justify-content: center;

      svg {
        width: 35px;

        path {
          fill: rgba(255, 255, 255, 1);
        }
      }
    }

    &:hover,
    &:focus {
      .toggle-bg {
        width: 100px;
        height: 65px;
        border-radius: 5px;
      }
    }
  }
`

export class Video extends React.Component {
  state = {
    videoIsPlaying: false,
  }
  videoWrapperRef = React.createRef()
  playVideo = () => {
    this.setState({ videoIsPlaying: true })
    setTimeout(() => {
      const videoWrapper = this.videoWrapperRef.current

      if (videoWrapper) {
        const video = videoWrapper.querySelector("iframe")

        if (video) {
          video.setAttribute("src", video.getAttribute("src") + "&autoplay=1")
        }
      }
    }, 0)
  }
  render() {
    const { className, style, ...props } = this.props
    return (
      <Wrapper
        {...{
          "data-testid": "component-video",
          className,
          style,
        }}
      >
        {this.state.videoIsPlaying ? (
          <div
            className="video-responsive"
            ref={this.videoWrapperRef}
            {...setHtml(props.video)}
          />
        ) : (
          <button className="video-toggle" onClick={this.playVideo}>
            <div className="toggle-bg" />
            <LazyLoadImage src={props.thumbnail} className="thumbnail" />
            <div className="icon-wrapper">
              <PlayIcon />
            </div>
          </button>
        )}

        {props.caption && <p className="caption">{props.caption}</p>}
      </Wrapper>
    )
  }
}

Video.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  video: PropTypes.string.isRequired,
  thumbnail: PropTypes.string.isRequired,
  caption: PropTypes.string,
}

Video.defaultProps = {
  className: "",
  style: {
    maxWidth: "720px",
    margin: "0 auto",
  },
}
