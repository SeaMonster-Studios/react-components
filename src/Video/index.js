import React from "react"
import PropTypes from "prop-types"
//
import { setHtml } from "../../utils"
import { LazyLoadImage } from "../LazyLoadImage"
import { Play as PlayIcon } from "../components/Icons"
import { Wrapper } from "./style"

export class Video extends React.Component {
  state = {
    videoIsPlaying: false,
  }
  videoWrapperRef = React.createRef()
  playVideo = () => {
    this.setState({ videoIsPlaying: true })
    setTimeout(() => {
      const videoWrapper = this.videoWrapperRef.current

      if (videoWrapper) {
        const video = videoWrapper.querySelector("iframe")

        if (video) {
          video.setAttribute("src", video.getAttribute("src") + "&autoplay=1")
        }
      }
    }, 0)
  }
  render() {
    const { className, style, ...props } = this.props
    return (
      <Wrapper
        {...{
          "data-testid": "component-video",
          className,
          style,
        }}
      >
        {this.state.videoIsPlaying ? (
          <div
            className="video-responsive"
            ref={this.videoWrapperRef}
            {...setHtml(props.video)}
          />
        ) : (
          <button className="video-toggle" onClick={this.playVideo}>
            <div className="toggle-bg" />
            <LazyLoadImage src={props.thumbnail} className="thumbnail" />
            <div className="icon-wrapper">
              <PlayIcon />
            </div>
          </button>
        )}

        {props.caption && <p className="caption">{props.caption}</p>}
      </Wrapper>
    )
  }
}

Video.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  video: PropTypes.string.isRequired,
  thumbnail: PropTypes.string.isRequired,
  caption: PropTypes.string,
}

Video.defaultProps = {
  className: "",
  style: {
    maxWidth: "720px",
    margin: "0 auto",
  },
}
