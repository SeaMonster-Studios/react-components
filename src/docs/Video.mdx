---
name: Video
---

import { Playground, PropsTable } from 'docz'
import { Video } from '../components/Video'
import faker from 'faker'

# Video

<PropsTable of={Video} />

<Playground>
  <Video caption={faker.lorem.paragraph()} thumbnail="https://www.valuecoders.com/blog/wp-content/uploads/2016/08/react.png" video={'<iframe width="500" height="281" src="https://www.youtube.com/embed/KVZ-P-ZI6W4?feature=oembed" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>'} />
</Playground>