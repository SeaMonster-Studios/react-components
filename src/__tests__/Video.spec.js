import React from "react"
import { render, cleanup, fireEvent, waitForElement } from "react-testing-library"
import { Video } from "../components/Video"

afterEach(cleanup)

describe("Video Component Test", () => {
  it("Renders in it's default state: Shows thumbnail and hides video", () => {
    const { getByTestId, queryByTestId, props, container } = renderSetup()
    const component = getByTestId("component-video")
    const video = queryByTestId("video")
    const thumbnail = container.querySelector(`img[src="${props.thumbnail}"]`)


    expect(video).toBeFalsy()
    expect(thumbnail).toBeTruthy()
    expect(component).toBeTruthy()
    expect(component).toMatchSnapshot()
  })

  it('Shows the video when the trigger button is clicked, and hides the thumbnail.', async () => {
    const { getByTestId, queryByTestId, props, container } = renderSetup()
    const trigger = getByTestId("trigger-button")

    await waitForElement(() => fireEvent.click(trigger))

    const thumbnail = container.querySelector(`img[src="${props.thumbnail}"]`)
    const video = queryByTestId("video")

    expect(video).toBeTruthy()
    expect(thumbnail).toBeFalsy()
  })
})

function renderSetup(overrides) {
  const props = {
    thumbnail:
      "https://www.valuecoders.com/blog/wp-content/uploads/2016/08/react.png",
    video:
      '<iframe width="500" height="281" src="https://www.youtube.com/embed/KVZ-P-ZI6W4?feature=oembed" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>',
    ...overrides,
  }

  const wrapper = render(<Video {...props} />)

  return {
    ...wrapper,
    props,
  }
}
