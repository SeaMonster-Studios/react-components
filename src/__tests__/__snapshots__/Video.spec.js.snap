// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Component Test Renders in it's default state: Shows thumbnail and hides video 1`] = `
<div
  class="css-1rhxlv5 e1tz5mhe0"
  data-testid="component-video"
  style="max-width: 720px; margin: 0px auto;"
>
  <button
    class="video-toggle"
    data-testid="trigger-button"
  >
    <div
      class="toggle-bg"
    />
    <img
      class="thumbnail"
      data-testid="component-lazy-load-image"
      src="https://www.valuecoders.com/blog/wp-content/uploads/2016/08/react.png"
    />
    <div
      class="icon-wrapper"
    >
      <svg
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 504c137 0 248-111 248-248S393 8 256 8 8 119 8 256s111 248 248 248zM40 256c0-118.7 96.1-216 216-216 118.7 0 216 96.1 216 216 0 118.7-96.1 216-216 216-118.7 0-216-96.1-216-216zm331.7-18l-176-107c-15.8-8.8-35.7 2.5-35.7 21v208c0 18.4 19.8 29.8 35.7 21l176-101c16.4-9.1 16.4-32.8 0-42zM192 335.8V176.9c0-4.7 5.1-7.6 9.1-5.1l134.5 81.7c3.9 2.4 3.8 8.1-.1 10.3L201 341c-4 2.3-9-.6-9-5.2z"
        />
      </svg>
    </div>
  </button>
</div>
`;
